src/lightning_app/runners/cloud.py

```diff
@@ -293,17 +287,91 @@ def dispatch(
         project = _get_project(self.backend.client)
 
         try:
-            list_apps_resp = self.backend.client.lightningapp_v2_service_list_lightningapps_v2(
-                project_id=project.project_id, name=app_config.name
+            if cluster_id is not None:
+                # Verify that the cluster exists
+                list_clusters_resp = self.backend.client.cluster_service_list_clusters()
+                cluster_ids = [cluster.id for cluster in list_clusters_resp.clusters]
+                if cluster_id not in cluster_ids:
+                    raise ValueError(f"You requested to run on cluster {cluster_id}, but that cluster doesn't exist.")
+
+                self._ensure_cluster_project_binding(project.project_id, cluster_id)
+
+            # Resolve the app name, instance, and cluster ID
+            existing_instance = None
+            app_name = app_config.name
+
+            # List existing instances
+            # TODO: Add pagination, otherwise this could break if users have a lot of apps.
+            find_instances_resp = self.backend.client.lightningapp_instance_service_list_lightningapp_instances(
+                project_id=project.project_id
             )
-            if list_apps_resp.lightningapps:
-                # There can be only one app with unique project_id<>name pair
-                lit_app = list_apps_resp.lightningapps[0]
-            else:
-                app_body = Body7(name=app_config.name, can_download_source_code=True)
+
+            # Seach for instances whose name starts with the given name
+            instances = [
+                lightningapp
+                for lightningapp in find_instances_resp.lightningapps
+                if lightningapp.name.startswith(app_name)
+            ]
+
+            # If instances exist and cluster is None, mimic cluster selection logic to choose a default
+            if cluster_id is None and len(instances) > 0:
+                # Determine the cluster ID
+                cluster_id = self._get_default_cluster(project.project_id)
+
+            # If an instance exists on the cluster with the base name - restart it
+            for instance in instances:
+                if instance.spec.cluster_id == cluster_id:
+                    existing_instance = instance
+                    break
+
+            # If instances exist but not on the cluster - choose a randomised name
+            if len(instances) > 0 and existing_instance is None:
```

Comment:
This is super confusing ? Why can't I have the same name on 2 different clusters ?